nomos-core/chain-defs/src/proofs/leader_proof.rs

```diff
@@ -1,100 +1,279 @@
-use nomos_proof_statements::leadership::{LeaderPrivate, LeaderPublic};
-use serde::{de::Error as _, Deserialize, Deserializer, Serialize, Serializer};
+use ark_ff::{BigInteger as _, Field as _, PrimeField as _};
+use groth16::{serde::serde_fr, Fr};
+use num_bigint::BigUint;
+use poseidon2::{Digest as _, Poseidon2Bn254Hasher};
+use serde::{Deserialize, Serialize};
 use thiserror::Error;
 
-#[derive(Debug, Clone)]
-pub struct Risc0LeaderProof {
-    public_inputs: LeaderPublic,
-    risc0_receipt: risc0_zkvm::Receipt,
+const POL_PROOF_DEV_MODE: &str = "POL_PROOF_DEV_MODE";
+
+use crate::{
+    mantle::{ledger::Utxo, ops::leader_claim::VoucherCm},
+    utils::merkle::{MerkleNode, MerklePath},
+};
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Groth16LeaderProof {
+    #[serde(with = "proof_serde")]
+    proof: pol::PoLProof,
+    #[serde(with = "serde_fr")]
+    entropy_contribution: Fr,
+    leader_key: ed25519_dalek::VerifyingKey,
+    voucher_cm: VoucherCm,
 }
 
 #[derive(Debug, Error)]
 pub enum Error {
-    #[error("Risc0 proof failed: {0}")]
-    Risc0ProofFailed(#[from] anyhow::Error),
+    #[error("Proof of leadership failed: {0}")]
+    PoLProofFailed(#[from] pol::ProveError),
 }
 
-impl Risc0LeaderProof {
-    pub fn prove(
-        public_inputs: LeaderPublic,
-        private_inputs: &LeaderPrivate,
-        prover: &dyn risc0_zkvm::Prover,
-    ) -> Result<Self, Error> {
-        let env = risc0_zkvm::ExecutorEnv::builder()
-            .write(&public_inputs)
-            .unwrap()
-            .write(&private_inputs)
-            .unwrap()
-            .build()
-            .unwrap();
-
+impl Groth16LeaderProof {
+    pub fn prove(private: &LeaderPrivate, voucher_cm: VoucherCm) -> Result<Self, Error> {
         let start_t = std::time::Instant::now();
+        let (proof, entropy_contribution) = if std::env::var(POL_PROOF_DEV_MODE).is_ok() {
+            tracing::warn!(
+                "Proofs are being generated in dev mode. This should never be used in production."
+            );
+            (
+                groth16::CompressedGroth16Proof::new(
+                    [0; 32].into(),
+                    [0; 64].into(),
+                    [0; 32].into(),
```

Comment:
Not sure if this will change as now they are `GenericArrray`, but I think `GenericArray::default()` would do (because it should know the sizes automatically).